<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  Top Level Namespace
  
    &mdash; Documentation by YARD 0.9.25
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index</a> &raquo;
    
    
    <span class="title">Top Level Namespace</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><h1>Top Level Namespace
  
  
  
</h1>
<div class="box_info">
  

  
  
  
  
  

  

  
</div>








  
    <h2>
      Instance Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#fetch-instance_method" title="#fetch (instance method)">#<strong>fetch</strong>(url, headers = {accept: &quot;*/*&quot;}, user = &quot;&quot;, pass = &quot;&quot;)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Obtain the information of an URL.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#get_AGI_Locus-instance_method" title="#get_AGI_Locus (instance method)">#<strong>get_AGI_Locus</strong>(path)  &#x21d2; locus_code </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Array with all the genes in the file.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#get_embl-instance_method" title="#get_embl (instance method)">#<strong>get_embl</strong>(all_locus)  &#x21d2; embl </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Hash with the name of the gene as key and Bio::EMBL object as value.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#get_exon-instance_method" title="#get_exon (instance method)">#<strong>get_exon</strong>(embl)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'></div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#get_gff3_chrom-instance_method" title="#get_gff3_chrom (instance method)">#<strong>get_gff3_chrom</strong>(new_embl)  &#x21d2; gff3_chrom </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>A GFF3 file.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#get_gff3_gene-instance_method" title="#get_gff3_gene (instance method)">#<strong>get_gff3_gene</strong>(new_embl)  &#x21d2; gff3_exon </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>A GFF3 file.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#report_no_CTT-instance_method" title="#report_no_CTT (instance method)">#<strong>report_no_CTT</strong>(new_embl)  &#x21d2; report </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>A report file.</p>
</div></span>
  
</li>

      
    </ul>
  



  <div id="instance_method_details" class="method_details_list">
    <h2>Instance Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="fetch-instance_method">
  
    #<strong>fetch</strong>(url, headers = {accept: &quot;*/*&quot;}, user = &quot;&quot;, pass = &quot;&quot;)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Obtain the information of an URL</p>


  </div>
</div>
<div class="tags">
  

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'CTTCTT_information.rb', line 11</span>

<span class='kw'>def</span> <span class='id identifier rubyid_fetch'>fetch</span><span class='lparen'>(</span><span class='id identifier rubyid_url'>url</span><span class='comma'>,</span> <span class='id identifier rubyid_headers'>headers</span> <span class='op'>=</span> <span class='lbrace'>{</span><span class='label'>accept:</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>*/*</span><span class='tstring_end'>&quot;</span></span><span class='rbrace'>}</span><span class='comma'>,</span> <span class='id identifier rubyid_user'>user</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_pass'>pass</span><span class='op'>=</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>Request</span><span class='period'>.</span><span class='id identifier rubyid_execute'>execute</span><span class='lparen'>(</span><span class='lbrace'>{</span>
    <span class='label'>method:</span> <span class='symbol'>:get</span><span class='comma'>,</span>
    <span class='label'>url:</span> <span class='id identifier rubyid_url'>url</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span><span class='comma'>,</span>
    <span class='label'>user:</span> <span class='id identifier rubyid_user'>user</span><span class='comma'>,</span>
    <span class='label'>password:</span> <span class='id identifier rubyid_pass'>pass</span><span class='comma'>,</span>
    <span class='label'>headers:</span> <span class='id identifier rubyid_headers'>headers</span><span class='rbrace'>}</span><span class='rparen'>)</span>
  <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>
  
  <span class='kw'>rescue</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>ExceptionWithResponse</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
    <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
    <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>  <span class='comment'># now we are returning &#39;False&#39;, and we will check that with an \&quot;if\&quot; statement in our main code
</span>  <span class='kw'>rescue</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
    <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
    <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>  <span class='comment'># now we are returning &#39;False&#39;, and we will check that with an \&quot;if\&quot; statement in our main code
</span>  <span class='kw'>rescue</span> <span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
    <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
    <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>  <span class='comment'># now we are returning &#39;False&#39;, and we will check that with an \&quot;if\&quot; statement in our main code
</span><span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_AGI_Locus-instance_method">
  
    #<strong>get_AGI_Locus</strong>(path)  &#x21d2; <tt>locus_code</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Returns array with all the genes in the file.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>path</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the path of the file where are the names of the genes</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>locus_code</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>array with all the genes in the file</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


38
39
40
41
42
43
44
45
46</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'CTTCTT_information.rb', line 38</span>

<span class='kw'>def</span> <span class='id identifier rubyid_get_AGI_Locus'>get_AGI_Locus</span><span class='lparen'>(</span><span class='id identifier rubyid_path'>path</span><span class='rparen'>)</span>
  <span class='id identifier rubyid_locus_code'>locus_code</span><span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>
  <span class='id identifier rubyid_locus'>locus</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='id identifier rubyid_path'>path</span><span class='comma'>,</span> <span class='label'>mode:</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>r</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_locus'>locus</span><span class='period'>.</span><span class='id identifier rubyid_readlines'>readlines</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_line'>line</span><span class='op'>|</span> 
    <span class='id identifier rubyid_code'>code</span> <span class='op'>=</span> <span class='id identifier rubyid_line'>line</span><span class='period'>.</span><span class='id identifier rubyid_strip'>strip</span><span class='period'>.</span><span class='id identifier rubyid_split'>split</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_locus_code'>locus_code</span> <span class='op'>|=</span> <span class='id identifier rubyid_code'>code</span>
  <span class='kw'>end</span> 
  <span class='kw'>return</span> <span class='id identifier rubyid_locus_code'>locus_code</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_embl-instance_method">
  
    #<strong>get_embl</strong>(all_locus)  &#x21d2; <tt>embl</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Returns hash with the name of the gene as key and Bio::EMBL object as
value.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>all_locus</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>array with genes</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>embl</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>hash with the name of the gene as key and Bio::EMBL object as value</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


54
55
56
57
58
59
60
61
62</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'CTTCTT_information.rb', line 54</span>

<span class='kw'>def</span> <span class='id identifier rubyid_get_embl'>get_embl</span><span class='lparen'>(</span><span class='id identifier rubyid_all_locus'>all_locus</span><span class='rparen'>)</span>
  <span class='id identifier rubyid_embl'>embl</span> <span class='op'>=</span> <span class='const'>Hash</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span>
  <span class='id identifier rubyid_all_locus'>all_locus</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_locus'>locus</span><span class='op'>|</span>
      <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='id identifier rubyid_fetch'><span class='object_link'><a href="#fetch-instance_method" title="#fetch (method)">fetch</a></span></span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>http://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&amp;format=embl&amp;id=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_locus'>locus</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
      <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_response'>response</span><span class='period'>.</span><span class='id identifier rubyid_nil?'>nil?</span>
      <span class='id identifier rubyid_embl'>embl</span><span class='lbracket'>[</span><span class='id identifier rubyid_locus'>locus</span><span class='rbracket'>]</span><span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>EMBL</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_response'>response</span><span class='period'>.</span><span class='id identifier rubyid_body'>body</span><span class='rparen'>)</span>
  <span class='kw'>end</span>
  <span class='kw'>return</span> <span class='id identifier rubyid_embl'>embl</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_exon-instance_method">
  
    #<strong>get_exon</strong>(embl)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    

  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>embl</span>
      
      
        <span class='type'>(<tt>Hash</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>hash with the name of the gene as key and Bio::EMBL object as value</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'CTTCTT_information.rb', line 69</span>

<span class='kw'>def</span> <span class='id identifier rubyid_get_exon'>get_exon</span><span class='lparen'>(</span><span class='id identifier rubyid_embl'>embl</span><span class='rparen'>)</span>
  
  <span class='id identifier rubyid_embl'>embl</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_key'>key</span><span class='comma'>,</span> <span class='id identifier rubyid_info'>info</span><span class='op'>|</span>
    <span class='id identifier rubyid_info'>info</span> <span class='op'>=</span> <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span> <span class='comment'>#we need to convert the database entry to a Bio::Sequence in order to add the new features
</span>     <span class='id identifier rubyid_all_starts_f'>all_starts_f</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span> <span class='comment'>#array to keep the forward locations without repetitions
</span>     <span class='id identifier rubyid_all_starts_r'>all_starts_r</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span> <span class='comment'>#array to keep the reverse locations without repetitions
</span>    
    <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
      <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>!=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>exon</span><span class='tstring_end'>&quot;</span></span>
      <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_location'>location</span><span class='op'>|</span>
      <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='op'>=</span><span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_subseq'>subseq</span><span class='lparen'>(</span><span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='comma'>,</span><span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span><span class='rparen'>)</span>
      <span class='id identifier rubyid_exon_start'>exon_start</span> <span class='op'>=</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span> <span class='comment'>#keep the start position of the exon in order to have then the location of the repeats in the gene
</span>        
      <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_nil?'>nil?</span>
        
        <span class='kw'>if</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='int'>1</span>
          <span class='comment'>#When the strand is positive we only have to search for the cttctt repeats
</span>          <span class='comment'>#In order to get the the overlapping matches I used lookahead assertion (?=)
</span>          <span class='comment'>#In order to get all the positions I used the code from this website, that generates an array with all start positions
</span>          <span class='comment'>#https://stackoverflow.com/questions/5241653/ruby-regex-match-and-get-positions-of/5241843#5241843
</span>          <span class='id identifier rubyid_start_f'>start_f</span> <span class='op'>=</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_enum_for'>enum_for</span><span class='lparen'>(</span><span class='symbol'>:scan</span><span class='comma'>,</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>(?=(cttctt))</span><span class='regexp_end'>/</span></span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span> <span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_last_match'>last_match</span><span class='period'>.</span><span class='id identifier rubyid_begin'>begin</span><span class='lparen'>(</span><span class='int'>0</span><span class='rparen'>)</span><span class='rbrace'>}</span> 
          <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_start_f'>start_f</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_nil?'>nil?</span>
          
          <span class='id identifier rubyid_start_f'>start_f</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_x'>x</span><span class='op'>|</span>
          <span class='comment'>#@Start_f has the positions starting in 0, but the feature position start in 1, we could sum +1 in order to get th same position.
</span>          <span class='comment'>#  Instead of sum 1 at the start_f positions I leave it like that and then when exon_start is added (start in 1) the problem is solved
</span>          <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_all_starts_f'>all_starts_f</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='id identifier rubyid_x'>x</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start'>exon_start</span><span class='rparen'>)</span> <span class='comment'>#skip repeated locations
</span>          <span class='id identifier rubyid_add_feature'>add_feature</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_x'>x</span><span class='op'>+</span><span class='id identifier rubyid_exon_start'>exon_start</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_x'>x</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start'>exon_start</span><span class='op'>+</span><span class='int'>5</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
          <span class='id identifier rubyid_add_feature'>add_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>repeat_motif</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
          <span class='id identifier rubyid_add_feature'>add_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>+</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
          <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_add_feature'>add_feature</span> 
            <span class='id identifier rubyid_all_starts_f'>all_starts_f</span><span class='period'>.</span><span class='id identifier rubyid_push'>push</span><span class='lparen'>(</span><span class='id identifier rubyid_x'>x</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start'>exon_start</span><span class='rparen'>)</span>
          <span class='kw'>end</span>
          
        <span class='kw'>end</span>
        
        <span class='kw'>if</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='op'>-</span><span class='int'>1</span>
          <span class='comment'>#@When the strand is negative we can search in the complement strand the cttctt repeats or we can search 
</span>          <span class='comment'>#  for the complement of cttctt  ( aagaag) in the positive strand 
</span>          <span class='comment'>#I used the second option, because I find it easier
</span>          <span class='comment'>#In order to get the the overlapping matches I used lookahead assertion (?=)
</span>          <span class='comment'>#In order to get all the positions I used the code from this website, that generates an array with all start positions
</span>          <span class='comment'>#https://stackoverflow.com/questions/5241653/ruby-regex-match-and-get-positions-of/5241843#5241843
</span>          <span class='id identifier rubyid_start_r'>start_r</span> <span class='op'>=</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_enum_for'>enum_for</span><span class='lparen'>(</span><span class='symbol'>:scan</span><span class='comma'>,</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>(?=(aagaag))</span><span class='regexp_end'>/</span></span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span> <span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_last_match'>last_match</span><span class='period'>.</span><span class='id identifier rubyid_begin'>begin</span><span class='lparen'>(</span><span class='int'>0</span><span class='rparen'>)</span><span class='rbrace'>}</span> <span class='comment'>#is not necessary to add 1 because when we sum the location of the exon is like sum 1 already
</span>          <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_start_r'>start_r</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_nil?'>nil?</span>
          
          <span class='id identifier rubyid_start_r'>start_r</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_x'>x</span><span class='op'>|</span>
          <span class='comment'>#@start_r has the positions starting in 0, but the feature position start in 1, we could sum +1 in order to get th same position.
</span>          <span class='comment'>#  Instead of sum 1 at the start_r positions I leave it like that and then when exon_start is added (start in 1) the problem is solved
</span>          <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_all_starts_r'>all_starts_r</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='id identifier rubyid_x'>x</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start'>exon_start</span><span class='rparen'>)</span> <span class='comment'>#skip repeated locations
</span>          <span class='id identifier rubyid_add_feature'>add_feature</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span><span class='id identifier rubyid_location'>location</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>complement(</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_x'>x</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start'>exon_start</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_x'>x</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start'>exon_start</span> <span class='op'>+</span><span class='int'>5</span><span class='embexpr_end'>}</span><span class='tstring_content'>)</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='comment'>#complement in order to indicate that they are in the negative strand
</span>          <span class='id identifier rubyid_add_feature'>add_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>repeat_motif</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
          <span class='id identifier rubyid_add_feature'>add_feature</span><span class='period'>.</span><span class='id identifier rubyid_append'>append</span><span class='lparen'>(</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>-</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='rparen'>)</span>
          <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_add_feature'>add_feature</span> 
            <span class='id identifier rubyid_all_starts_r'>all_starts_r</span><span class='period'>.</span><span class='id identifier rubyid_push'>push</span><span class='lparen'>(</span><span class='id identifier rubyid_x'>x</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start'>exon_start</span><span class='rparen'>)</span>
          <span class='kw'>end</span>
        <span class='kw'>end</span>
     <span class='kw'>end</span>
    <span class='kw'>end</span>
  <span class='kw'>end</span>

<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_gff3_chrom-instance_method">
  
    #<strong>get_gff3_chrom</strong>(new_embl)  &#x21d2; <tt>gff3_chrom</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Returns a GFF3 file.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>new_embl</span>
      
      
        <span class='type'>(<tt>Hash</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>hash with the name of the gene as key and Bio::Sequence object with new
features added</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>gff3_chrom</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>a GFF3 file</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'CTTCTT_information.rb', line 184</span>

<span class='kw'>def</span> <span class='id identifier rubyid_get_gff3_chrom'>get_gff3_chrom</span><span class='lparen'>(</span><span class='id identifier rubyid_new_embl'>new_embl</span><span class='rparen'>)</span>
<span class='comment'>#seqid - ID of the chromosome
</span><span class='comment'>#type - type of feature. Must be a term or accession from the SOFA sequence ontology
</span><span class='comment'>#start - Start position of the feature, with sequence numbering starting at 1.
</span><span class='comment'>#end - End position of the feature, with sequence numbering starting at 1.
</span><span class='comment'>#score - A floating point value.
</span><span class='comment'>#strand - defined as + (forward) or - (reverse).
</span><span class='comment'>#phase - . 
</span><span class='comment'>#attributes - A semicolon-separated list of tag-value pairs, providing additional information.
</span>  <span class='id identifier rubyid_gff3_chrom'>gff3_chrom</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>gff3_chrom</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='label'>mode:</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>w</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_gff3_chrom'>gff3_chrom</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>##gff-version 3\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='comment'>#the first line of the gff3 file
</span>  <span class='id identifier rubyid_type'>type</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>SO:0000657</span><span class='tstring_end'>&quot;</span></span>
  <span class='id identifier rubyid_source'>source</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Ensembl</span><span class='tstring_end'>&quot;</span></span>
  <span class='id identifier rubyid_score'>score</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  <span class='id identifier rubyid_phase'>phase</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  
  <span class='id identifier rubyid_new_embl'>new_embl</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_key'>key</span><span class='comma'>,</span> <span class='id identifier rubyid_info'>info</span><span class='op'>|</span>
    <span class='id identifier rubyid_info'>info</span> <span class='op'>=</span> <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span>
    <span class='id identifier rubyid_count'>count</span> <span class='op'>=</span> <span class='int'>0</span>
    <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
      <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>!=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&quot;</span></span>
      <span class='id identifier rubyid_count'>count</span> <span class='op'>+=</span> <span class='int'>1</span>
      <span class='id identifier rubyid_seqid'>seqid</span> <span class='op'>=</span> <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_primary_accession'>primary_accession</span><span class='period'>.</span><span class='id identifier rubyid_split'>split</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>:</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='lbracket'>[</span><span class='int'>2</span><span class='rbracket'>]</span> <span class='comment'>#the number of the chromosome
</span>      <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_location'>location</span><span class='op'>|</span>
      <span class='id identifier rubyid_chr_start'>chr_start</span> <span class='op'>=</span> <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_primary_accession'>primary_accession</span><span class='period'>.</span><span class='id identifier rubyid_split'>split</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>:</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='lbracket'>[</span><span class='int'>3</span><span class='rbracket'>]</span> <span class='comment'>#start position of the chromosome
</span>      
      <span class='comment'>#@We add to the positions the start position of the chromosome.
</span>      <span class='comment'>#  Both positions start in 1 so we need to rest 1 in order to obtain the correct position
</span>      <span class='id identifier rubyid_start'>start</span><span class='op'>=</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>+</span> <span class='id identifier rubyid_chr_start'>chr_start</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>-</span> <span class='int'>1</span>
      <span class='id identifier rubyid_end_'>end_</span><span class='op'>=</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span>  <span class='op'>+</span> <span class='id identifier rubyid_chr_start'>chr_start</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>-</span> <span class='int'>1</span>
      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span>
      <span class='id identifier rubyid_attributes'>attributes</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='period'>.</span><span class='id identifier rubyid_keys'>keys</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>_</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_count'>count</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='embexpr_end'>}</span><span class='tstring_content'>;Name=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span><span class='embexpr_end'>}</span><span class='tstring_content'>;</span><span class='tstring_end'>&quot;</span></span>
      
      <span class='id identifier rubyid_gff3_chrom'>gff3_chrom</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqid'>seqid</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_source'>source</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_type'>type</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start'>start</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_'>end_</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_score'>score</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_phase'>phase</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_attributes'>attributes</span><span class='embexpr_end'>}</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
        
      <span class='kw'>end</span>
    <span class='kw'>end</span> 
  <span class='kw'>end</span>
  <span class='id identifier rubyid_gff3_chrom'>gff3_chrom</span><span class='period'>.</span><span class='id identifier rubyid_close'>close</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_gff3_gene-instance_method">
  
    #<strong>get_gff3_gene</strong>(new_embl)  &#x21d2; <tt>gff3_exon</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Returns a GFF3 file.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>new_embl</span>
      
      
        <span class='type'>(<tt>Hash</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>hash with the name of the gene as key and Bio::Sequence object with new
features added</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>gff3_exon</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>a GFF3 file</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'CTTCTT_information.rb', line 142</span>

<span class='kw'>def</span> <span class='id identifier rubyid_get_gff3_gene'>get_gff3_gene</span><span class='lparen'>(</span><span class='id identifier rubyid_new_embl'>new_embl</span><span class='rparen'>)</span>
  
<span class='comment'>#seqid - ID of the gene
</span><span class='comment'>#type - type of feature. Must be a term or accession from the SOFA sequence ontology
</span><span class='comment'>#start - Start position of the feature, with sequence numbering starting at 1.
</span><span class='comment'>#end - End position of the feature, with sequence numbering starting at 1.
</span><span class='comment'>#score - A floating point value.
</span><span class='comment'>#strand - defined as + (forward) or - (reverse).
</span><span class='comment'>#phase - . 
</span><span class='comment'>#attributes - A semicolon-separated list of tag-value pairs, providing additional information.
</span>  <span class='id identifier rubyid_gff3_exon'>gff3_exon</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>gff3_exon</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='label'>mode:</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>w</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span>
  <span class='id identifier rubyid_gff3_exon'>gff3_exon</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>##gff-version 3\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='comment'>#the first line of the gff3 file
</span>  <span class='id identifier rubyid_type'>type</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>SO:0000657</span><span class='tstring_end'>&quot;</span></span> <span class='comment'>#Sequence Ontology repeat-region
</span>  <span class='id identifier rubyid_source'>source</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Ensembl</span><span class='tstring_end'>&quot;</span></span>
  <span class='id identifier rubyid_score'>score</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  <span class='id identifier rubyid_phase'>phase</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
  
  <span class='id identifier rubyid_new_embl'>new_embl</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_key'>key</span><span class='comma'>,</span> <span class='id identifier rubyid_info'>info</span><span class='op'>|</span>
    <span class='id identifier rubyid_info'>info</span> <span class='op'>=</span> <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span>
    <span class='id identifier rubyid_count'>count</span> <span class='op'>=</span> <span class='int'>0</span>
    <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
      <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>!=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&quot;</span></span>
      <span class='id identifier rubyid_count'>count</span> <span class='op'>+=</span> <span class='int'>1</span>
      <span class='id identifier rubyid_seqid'>seqid</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_key'>key</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span> <span class='comment'>#this is the seqID used in Ensembl, so I used the same name
</span>      <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_location'>location</span><span class='op'>|</span>
      <span class='id identifier rubyid_start'>start</span><span class='op'>=</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span>
      <span class='id identifier rubyid_end_'>end_</span><span class='op'>=</span> <span class='id identifier rubyid_location'>location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span>  
      <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span>
      <span class='id identifier rubyid_attributes'>attributes</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='period'>.</span><span class='id identifier rubyid_keys'>keys</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>_</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_count'>count</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='embexpr_end'>}</span><span class='tstring_content'>;Name=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span><span class='embexpr_end'>}</span><span class='tstring_content'>;</span><span class='tstring_end'>&quot;</span></span>
      
      <span class='id identifier rubyid_gff3_exon'>gff3_exon</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_seqid'>seqid</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_source'>source</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_type'>type</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start'>start</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_'>end_</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_score'>score</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_phase'>phase</span><span class='embexpr_end'>}</span><span class='tstring_content'>\t</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_attributes'>attributes</span><span class='embexpr_end'>}</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
        
      <span class='kw'>end</span>
    <span class='kw'>end</span> 
  <span class='kw'>end</span>
  <span class='id identifier rubyid_gff3_exon'>gff3_exon</span><span class='period'>.</span><span class='id identifier rubyid_close'>close</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="report_no_CTT-instance_method">
  
    #<strong>report_no_CTT</strong>(new_embl)  &#x21d2; <tt>report</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Returns a report file.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>new_embl</span>
      
      
        <span class='type'>(<tt>Hash</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>hash with the name of the gene as key and Bio::Sequence object with new
features added</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>report</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>a report file</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'CTTCTT_information.rb', line 229</span>

<span class='kw'>def</span> <span class='id identifier rubyid_report_no_CTT'>report_no_CTT</span><span class='lparen'>(</span><span class='id identifier rubyid_new_embl'>new_embl</span><span class='rparen'>)</span>
  
  <span class='id identifier rubyid_report'>report</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_open'>open</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>report_no_CTT</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='label'>mode:</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>w</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span>
      <span class='id identifier rubyid_report'>report</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>*******************************************************************************\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
      <span class='id identifier rubyid_report'>report</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>*******************************REPORT FILE*************************************\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
      <span class='id identifier rubyid_report'>report</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>*******************************************************************************\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
      <span class='id identifier rubyid_report'>report</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>*************These are the gene on the list without CTT repeats**************\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
   <span class='id identifier rubyid_new_embl'>new_embl</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_key'>key</span><span class='comma'>,</span> <span class='id identifier rubyid_info'>info</span><span class='op'>|</span>
    <span class='id identifier rubyid_info'>info</span> <span class='op'>=</span> <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span>
    <span class='id identifier rubyid_count'>count</span> <span class='op'>=</span> <span class='int'>0</span>
    <span class='id identifier rubyid_info'>info</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
      <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>!=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&quot;</span></span>
      <span class='id identifier rubyid_count'>count</span> <span class='op'>+=</span> <span class='int'>1</span>
    <span class='kw'>end</span>
       <span class='kw'>if</span> <span class='id identifier rubyid_count'>count</span> <span class='op'>==</span> <span class='int'>0</span> <span class='comment'>#when the count is 0, we add the key of the gene, because that means that it have 0 CTT repeats
</span>         <span class='id identifier rubyid_report'>report</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_key'>key</span><span class='embexpr_end'>}</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
       <span class='kw'>end</span>
    <span class='kw'>end</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

</div>

      <div id="footer">
  Generated on Sun Nov 28 04:32:03 2021 by
  <a href="http://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.25 (ruby-2.5.1).
</div>

    </div>
  </body>
</html>